File: Data Structures & Algorithms/count-number-of-islands/submission-3.js
class Solution {
    dfs(i,j,grid){
        if(i<0 || j<0 || i>= grid.length || j>= grid[0].length || grid[i][j] =='0')
            return ;
        grid[i][j]='0';
        this.dfs(i+1,j,grid);
        this.dfs(i-1,j,grid);
        this.dfs(i,j+1,grid);
        this.dfs(i,j-1,grid);
    }
    numIslands(grid) {
        let count =0 ;
        for(let i=0;i<grid.length ; i++){
            for(let j=0 ; j<grid[i].length;j++){
                if(grid[i][j]=='1'){
                    this.dfs(i,j,grid);
                    count++;
                }
            }
        }
        return count;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-3.cpp
class Solution {
public:
    set<pair<int,int>> s ;
    bool dfs(int ligne, int col,int i,vector<vector<char>>& board, string word){
        if(i==word.size()) return true;
        if(ligne<0 || col<0 || ligne >= board.size() || col >= board[0].size() ||
        s.count({ligne,col}) || word[i] != board[ligne][col] ) return false;
        s.insert({ligne,col});
        bool res= (dfs(ligne-1,col,i+1,board,word) || dfs(ligne+1,col,i+1,board,word) ||
        dfs(ligne,col-1,i+1,board,word) || dfs(ligne,col+1,i+1,board,word) );
        s.erase({ligne,col});
        return res;
    }
    bool exist(vector<vector<char>>& board, string word) {
        for(int i=0;i<board.size();i++)
            for(int j=0;j<board[0].size();j++)
                if(dfs(i,j,0,board,word)) return true;
        return false;
    }
};
